File: Array/SlidingWindow/minWindow.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        need, missing = collections.Counter(t), len(t)
        left = 0
        minLeft = 0
        minRight = 0
        for right, c in enumerate(s, 1):
            missing -= need[c] > 0
            need[c] -= 1
            if not missing:
                while left < right and need[s[left]] < 0:
                    need[s[left]] += 1
                    left += 1
                if not minRight or right - left <= minRight - minLeft:
                    minLeft = left
                    minRight= right
        return s[minLeft:minRight]

File: Array/SlidingWindow/minWindow.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        if (t.length() > s.length()) return "";
        unordered_map<char, int> needs;
        int missing = t.length();
        
        int left = 0; int minLeft = 0; int minRight = 0;
        
        for (auto &ch : s) {
            if (needs.find(ch) != needs.end()) continue;
            else needs.insert({ch, 0});
        }
        
        for (auto &ch : t) {
            if (needs.find(ch) != needs.end()) needs[ch] ++;
            else needs.insert({ch, 1});
            
        }
        for (int right = 0; right < s.length(); ++right) {
            missing -= (needs[s[right]] > 0);
            needs[s[right]] --;
            
            if (!missing) {
                while (left <= right && 
                       needs[s[left]] < 0) {
                    needs[s[left]] ++;
                    left ++;
                }
                if (!minRight || right - left + 1 <= minRight - minLeft) {
                    minRight = right+1;
                    minLeft = left;
                }
            }
        }

        return s.substr(minLeft, minRight - minLeft);
    }
};